File: 0094-binary-tree-inorder-traversal.py
"""
94. Binary Tree Inorder Traversal

Submitted: March 2, 2023

Runtime: 32 ms (beats 100.00%)
Memory: 13.92 MB (beats 100.00%)
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        f=[]
        def z(n):
            if n is None:
                return
            z(n.left)
            f.append(n.val)
            z(n.right)
        z(root)
        return f

File: 0094-binary-tree-inorder-traversal.cpp
/*
94. Binary Tree Inorder Traversal

Submitted: December 2, 2024

Runtime: 0 ms (beats 100.00%)

Memory: 11.04 MB (beats 14.19%)
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        traversalHelper(result, root);
        return result;
    }
    void traversalHelper(vector<int>& vec, TreeNode* root) {
        if (root == nullptr) return;
        traversalHelper(vec, root->left);
        vec.push_back(root->val);
        traversalHelper(vec, root->right);
    }
};
